### PR TITLE
libclc: increase fp16 support

### PR DESCRIPTION
**Pull request by rjodinchr**
Increase fp16 support to allow clspv to continue to be OpenCL compliant following the update of the OpenCL-CTS adding more testing on math functions and conversions with half.

Math functions are implemented by upscaling to fp32 and using the fp32 implementation. It garantees the accuracy required for half-precision float-point by the CTS.

### CODE REVIEW

**Comment by rjodinchr** [comment]
Those implementations have been tested using 1 Intel and 1 AMD devices (with clvk/clspv)

**Comment by rjodinchr** [comment]
@alan-baker @kpet , could you please review?

**Review by alan-baker — APPROVED** [review]
Looks reasonable to me.

**Comment by rjodinchr** [comment]
@frasercrmck Could you review it please?

**Comment by frasercrmck** [comment]
> @frasercrmck Could you review it please?

Sure, will do, thanks for this. We've done something similar downstream so I can hopefully verify against what we've done.

File: libclc/generic/lib/math/atan.cl [hunk 1]
```diff
@@ -181,3 +181,15 @@ _CLC_OVERLOAD _CLC_DEF double atan(double x)
 _CLC_UNARY_VECTORIZE(_CLC_OVERLOAD _CLC_DEF, double, atan, double);
 
 #endif // cl_khr_fp64
+
+#ifdef cl_khr_fp16
```
Comment (frasercrmck) [inline libclc/generic/lib/math/atan.cl:185]: Can this be `_CLC_DEFINE_UNARY_BUILTIN_FP16(atan)`?

**Review by frasercrmck — APPROVED** [review]
LGTM, thanks

File: libclc/generic/lib/math/atan.cl (hunk 1 above)
Comment (rjodinchr) [reply libclc/generic/lib/math/atan.cl:185]: I agree, but I had some weird compilation errors. Does it work on your side?

Comment (rjodinchr) [reply libclc/generic/lib/math/atan.cl:185]: Alright, just found that it is due to includes order in `atan.cl`. I'll change that

**Comment by rjodinchr** [comment]
Also, just leave that thought here: A next step might be to use `native` functions instead (when available). They should provide enough precision for half, but be more performance. Also as the accuracy is undefined for those functions, maybe it should be an opt-in option in the libclc (clspv would enable it for sure).

**Comment by rjodinchr** [comment]
@frasercrmck could you submit that PR? I cannot do it.

**Comment by frasercrmck** [comment]
> Also, just leave that thought here: A next step might be to use `native` functions instead (when available). They should provide enough precision for half, but be more performance. Also as the accuracy is undefined for those functions, maybe it should be an opt-in option in the libclc (clspv would enable it for sure).

Yeah, it would definitely need to be opt-in. Ideally we'd also implement the `half` builtins properly so they're not going through `float` and are as fast as they can be, but then _also_ allow native functions if users so wish. But that involves maths, haha.

**Comment by rjodinchr** [comment]
> @rjodinchr I was about to merge and noticed your email is hidden. I just wanted to check with you whether you want to make it public, as per the discussion https://discourse.llvm.org/t/hidden-emails-on-github-should-we-do-something-about-it/74223/67. AFAICT there's still no official policy, but thought I'd check.

I was not aware my email was private. It should be public now. Thanks

**Comment by rjodinchr** [comment]
I am able to reproduce it, working on it right now

**Comment by rjodinchr** [comment]
This PR should fix the build issue: https://github.com/llvm/llvm-project/pull/99481

**Comment by mgorny** [comment]
This change seems to have caused test regression. This is Gentoo Linux amd64, tried LLVM 72d8c2737bb557af9d0c735b9fa30b1b03485627 and 18.1.8.

To reproduce, I've done:

```
cmake -G Ninja "${llvm_project}"/libclc -DLIBCLC_TARGETS_TO_BUILD="spirv-mesa3d-;spirv64-mesa3d-;r600--;amdgcn--;amdgcn-mesa-mesa3d;amdgcn--amdhsa"
ninja
ninja test
```

The result:

```
[0/1] Running tests...
Test project /home/mgorny/git/llvm-project/libclc/build
    Start 1: external-calls-tahiti-amdgcn--.bc
1/7 Test #1: external-calls-tahiti-amdgcn--.bc .............***Failed    5.46 sec
    Start 2: external-calls-amdgcn--amdhsa.bc
2/7 Test #2: external-calls-amdgcn--amdhsa.bc ..............***Failed    5.54 sec
    Start 3: external-calls-tahiti-amdgcn-mesa-mesa3d.bc
3/7 Test #3: external-calls-tahiti-amdgcn-mesa-mesa3d.bc ...***Failed    5.43 sec
    Start 4: external-calls-cedar-r600--.bc
4/7 Test #4: external-calls-cedar-r600--.bc ................   Passed    2.05 sec
    Start 5: external-calls-cypress-r600--.bc
5/7 Test #5: external-calls-cypress-r600--.bc ..............   Passed    2.07 sec
    Start 6: external-calls-barts-r600--.bc
6/7 Test #6: external-calls-barts-r600--.bc ................   Passed    2.03 sec
    Start 7: external-calls-cayman-r600--.bc
7/7 Test #7: external-calls-cayman-r600--.bc ...............   Passed    2.07 sec

57% tests passed, 3 tests failed out of 7

Total Test time (real) =  24.66 sec

The following tests FAILED:
	  1 - external-calls-tahiti-amdgcn--.bc (Failed)
	  2 - external-calls-amdgcn--amdhsa.bc (Failed)
	  3 - external-calls-tahiti-amdgcn-mesa-mesa3d.bc (Failed)
Errors while running CTest
Output from these tests are in: /home/mgorny/git/llvm-project/libclc/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
FAILED: CMakeFiles/test.util 
cd /home/mgorny/git/llvm-project/libclc/build && /usr/bin/ctest --force-new-ctest-process
ninja: build stopped: subcommand failed.
```

LastTest.log: [LastTest.log](https://github.com/user-attachments/files/16320006/LastTest.log)

If I go back to 6838c7afb2d2fa5eb642a4b5922ae55d97f7b344, they all pass.

**Comment by rjodinchr** [comment]
@mgorny thank you for the report. It should be fixed with: https://github.com/llvm/llvm-project/pull/99841

**Pull request merged**